File: src/main/java/org/gjt/jclasslib/io/ClassFileConsistencyTest.java
/*
 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public
 License as published by the Free Software Foundation; either
 version 2 of the license, or (at your option) any later version.
 */

package org.gjt.jclasslib.io;

import org.gjt.jclasslib.structures.ClassFile;
import org.gjt.jclasslib.structures.InvalidByteCodeException;

import java.io.*;
import java.util.Enumeration;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;

public class ClassFileConsistencyTest {

    private static final ThreadLocal<byte[]> BUFFER_TL = new ThreadLocal<byte[]>() {
        @Override
        protected byte[] initialValue() {
            return new byte[1024];
        }
    };

    private static void scanJar(File file) throws IOException, InvalidByteCodeException {
        JarFile jar = new JarFile(file);
        Enumeration<JarEntry> entries = jar.entries();
        int count = 0;
        int errors = 0;
        while (entries.hasMoreElements()) {
            JarEntry entry = entries.nextElement();
            String name = entry.getName();
            int index = name.lastIndexOf(".class");
            if (index > 0) {
                String className = name.substring(0, index).replace("/", ".");
                try {
                    if (!checkClassFile(className, jar, entry)) {
                        errors++;
                    }
                } catch (IOException e) {
                    error(className);
                    throw e;
                } catch (InvalidByteCodeException e) {
                    error(className);
                    throw e;
                } catch (RuntimeException e) {
                    error(className);
                    throw e;
                }
                count++;
            }
        }
        System.out.println(String.valueOf(count) + " classes checked, " + errors + " errors");
    }

    private static void error(String className) {
        System.err.println("ERROR when processing " + className);
    }

    private static boolean checkClassFile(String className, JarFile jar, JarEntry entry) throws IOException, InvalidByteCodeException {

        ByteArrayOutputStream os = new ByteArrayOutputStream();
        InputStream is = jar.getInputStream(entry);
        pumpStream(is, os);
        is.close();
        byte[] before = os.toByteArray();

        ClassFile classFile = ClassFileReader.readFromInputStream(jar.getInputStream(entry));
        byte[] after = ClassFileWriter.writeToByteArray(classFile);

        boolean success = compare(className, before, after);
        if (!success) {
            ClassFileReader.readFromInputStream(new ByteArrayInputStream(after));
        }
        return success;
    }

    private static boolean compare(String className, byte[] before, byte[] after) {
        if (before.length != after.length) {
            System.err.println("ERROR in " + className);
            System.err.println("Different length " + before.length + " != " + after.length);
            return false;
        }

        for (int i = 0; i < before.length; i++) {
            if (before[i] != after[i]) {
                System.err.println("Different byte at index " + i);
                return false;
            }
        }

        return true;
    }

    public static long pumpStream(InputStream is, OutputStream os) throws IOException {

        byte[] buffer = BUFFER_TL.get();
        int count;
        while ((count = is.read(buffer)) != -1) {
            os.write(buffer, 0, count);
        }
        return count;
    }

    public static void main(String[] args) throws Exception {
        String javaHome = System.getProperty("java.home");
        File rtJar = new File(javaHome + File.separator + "lib" + File.separator + "rt.jar");
        scanJar(rtJar);
    }
}

File: modules/data/src/test/kotlin/test/ClassFileConsistencyTest.kt
/*
 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public
 License as published by the Free Software Foundation; either
 version 2 of the license, or (at your option) any later version.
 */

package org.gjt.jclasslib.test

import org.gjt.jclasslib.io.ClassFileReader
import org.gjt.jclasslib.io.ClassFileWriter
import org.gjt.jclasslib.io.forEachClassInJrt
import org.gjt.jclasslib.structures.isDebug
import org.testng.annotations.Test
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.net.URL
import java.nio.file.Path
import java.util.jar.JarFile

class ClassFileConsistencyTest {
    @Test
    fun testJres() {
        withAllJres { javaHome ->
            scanJre(javaHome)
        }
    }

    @Test
    fun testJreModules() {
        withAllJres { javaHome ->
            scanJreModules(javaHome)
        }
    }

    @Test
    fun testSingleClasses() {
        checkClassFile("/Main.class")
        checkClassFile("/moduleMainClass/module-info.class")
    }

    private fun scanJre(javaHome: File) {
        val testStatistics = TestStatistics()
        val rtJar = javaHome.resolve("lib/rt.jar")
        if (rtJar.exists()) {
            scanJar(rtJar, testStatistics)
        } else {
            scanJrt(javaHome, testStatistics)
        }
        println("${testStatistics.count} classes checked, ${testStatistics.errors} errors")
    }

    private fun scanJreModules(javaHome: File, logPaths: Boolean = false) {
        val rtJar = javaHome.resolve("lib/rt.jar")
        if (rtJar.exists()) {
            println("Not a modular JRE")
            return
        }
        val testStatistics = TestStatistics()
        scanJrt(javaHome, testStatistics, logPaths) { path ->
            path.endsWith("module-info.class")
        }
        println("${testStatistics.count} modules checked, ${testStatistics.errors} errors")
    }

    private fun scanJar(file: File, testStatistics: TestStatistics) {
        val jar = JarFile(file)
        for (entry in jar.entries().iterator()) {
            val fileName = entry.name
            if (fileName.endsWith(".class")) {
                checkClassFile(fileName, JarInputStreamProvider(jar, entry), testStatistics)
            }
        }
    }

    private fun scanJrt(jreHome: File, testStatistics: TestStatistics, logPaths: Boolean = false, pathFilter: (Path) -> Boolean = { true }) {
        forEachClassInJrt(jreHome) { path ->
            if (pathFilter(path)) {
                if (logPaths) {
                    println("Processing $path")
                }
                val fileName = path.toString()
                checkClassFile(fileName, JrtInputStreamProvider(fileName, jreHome), testStatistics)
            }
        }
    }

    private data class TestStatistics(var count: Int = 0, var errors: Int = 0)

    private fun checkClassFile(fileName: String, inputStreamProvider: InputStreamProvider, testStatistics: TestStatistics) {
        val className = fileName.toClassNameWithModuleCheck()
        try {
            if (!checkClassFile(className, inputStreamProvider)) {
                testStatistics.errors++
            }
        } catch (e: Throwable) {
            error(className)
            throw e
        }
        testStatistics.count++
    }

    private fun withAllJres(block: (File) -> Unit) {
        for (javaHome in getJavaHomes()) {
            println("Testing with $javaHome")
            block(javaHome)
        }
    }

    private fun getJavaHomes(): List<File> {
        val majorVersions = System.getProperty("majorVersions", "")
                .split(",")
                .mapNotNull { it.toIntOrNull() }
        val javaHomes = majorVersions.map {
            requireNotNull(System.getProperty("javaHome.$it")) {
                "Java $it home not specified"
            }
        }.ifEmpty { listOf(System.getProperty("java.home")) }.map { getJavaHome(it) }
        return javaHomes
    }

    private fun getJavaHome(javaHomePath: String): File {
        val javaHome = File(javaHomePath)
        val jreHome = javaHome.resolve("jre")
        return if (jreHome.exists()) {
            jreHome
        } else {
            javaHome
        }
    }

    private fun String.toClassNameWithModuleCheck(): String {
        val matchResult = Regex("modules/(.*?)/(.*)").find(this)
        return if (matchResult != null) {
            matchResult.groupValues[1] + "/" + matchResult.groupValues[2].toClassName()
        } else {
            toClassName()
        }
    }

    private fun String.toClassName(): String {
        return removeSuffix(".class").replace("/", ".")
    }

    private fun checkClassFile(resourcePath: String): Boolean {
        return checkClassFile(requireNotNull(InputStreamProvider::class.java.getResource(resourcePath)))
    }

    private fun checkClassFile(url: URL): Boolean {
        return checkClassFile(null, UrlInputStreamProvider(url))
    }

    private fun checkClassFile(className: String?, inputStreamProvider: InputStreamProvider): Boolean {
        val output = ByteArrayOutputStream()
        inputStreamProvider.createInputStream().use {
            it.copyTo(output)
        }
        val before = output.toByteArray()

        val classFile = ClassFileReader.readFromInputStream(inputStreamProvider.createInputStream())
        val after = ClassFileWriter.writeToByteArray(classFile)
        val reportedClassName = className ?: classFile.thisClassName.replace('/', '.')

        val success = compare(reportedClassName, before, after)
        if (!success) {
            isDebug = true
            println()
            println("*** before: ")
            ClassFileReader.readFromInputStream(ByteArrayInputStream(before))
            println()
            println("*** write: ")
            ClassFileWriter.writeToByteArray(classFile)
            println()
            println("*** after: ")
            ClassFileReader.readFromInputStream(ByteArrayInputStream(after))
            isDebug = false
        }
        return success
    }

    private fun error(className: String) {
        println("ERROR when processing $className")
    }

    private fun compare(className: String, before: ByteArray, after: ByteArray): Boolean {
        if (before.size != after.size) {
            println("ERROR in $className")
            println("Different length " + before.size + " != " + after.size)
        }
        for (i in 0 until minOf(before.size, after.size)) {
            if (before[i] != after[i]) {
                println("ERROR in $className")
                println("Different byte at index $i")
                println("" + before[i] + " != " + after[i])
                return false
            }
        }
        return true
    }
}
